File: ABS/Ruby/ABC049C.rb
wds = ['dream', 'dreamer', 'erase', 'eraser']
s = gets.strip

until s.empty? do
  isr = false
  wds.map do |w|
    if s[-w.size, w.size]==w
      s = s[0..-w.size-1]
      isr = true
      break
    end
  end

  unless isr
    puts 'NO'
    exit 0
  end
end

puts 'YES'

File: ABS/JavaScript/ABC049C.js
function main(input) {
    const words = ['dream', 'dreamer', 'erase', 'eraser'];
    const S = input.split('\n')[0];
    var i=S.length,w;

    while(true) {
        w = words.find((w) => S.endsWith(w, i));
        if(!w)
            break;
        i -= w.length;
    }
    
    console.log((i===0) ? 'YES' : 'NO');
 }
 
 main(require('fs').readFileSync('/dev/stdin', 'utf8'));